cnb_index.go

```diff
@@ -0,0 +1,1254 @@
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrOSUndefined = func(format types.MediaType, digest string) error {
+		return fmt.Errorf("image os is undefined for %s ImageIndex (digest: %s)", indexMediaType(format), digest)
+	}
+	ErrArchUndefined = func(format types.MediaType, digest string) error {
+		return fmt.Errorf("image architecture is undefined for %s ImageIndex (digest: %s)", indexMediaType(format), digest)
+	}
+	ErrVariantUndefined = func(format types.MediaType, digest string) error {
+		return fmt.Errorf("image variant is undefined for %s ImageIndex (digest: %s)", indexMediaType(format), digest)
+	}
+	ErrOSVersionUndefined = func(format types.MediaType, digest string) error {
+		return fmt.Errorf("image os-version is undefined for %s ImageIndex (digest: %s)", indexMediaType(format), digest)
+	}
+	ErrFeaturesUndefined = func(format types.MediaType, digest string) error {
+		return fmt.Errorf("image features is undefined for %s ImageIndex (digest: %s)", indexMediaType(format), digest)
+	}
+	ErrOSFeaturesUndefined = func(format types.MediaType, digest string) error {
+		return fmt.Errorf("image os-features is undefined for %s ImageIndex (digest: %s)", indexMediaType(format), digest)
+	}
+	ErrAnnotationsUndefined = func(format types.MediaType, digest string) error {
+		return fmt.Errorf("image annotations is undefined for %s ImageIndex (digest: %s)", indexMediaType(format), digest)
+	}
+	ErrNoImageOrIndexFoundWithGivenDigest = func(digest string) error {
+		return fmt.Errorf(`no image or image index found for digest "%s"`, digest)
+	}
+	ErrManifestUndefined   = errors.New("encountered unexpected error while parsing image: manifest or index manifest is nil")
+	ErrPlatformUndefined   = errors.New("unable to determine image platform: platform is nil")
+	ErrInvalidPlatform     = errors.New("unable to determine image platform: platform's 'OS' or 'Architecture' field is nil")
+	ErrConfigFileUndefined = errors.New("unable to access image configuration: ConfigFile is nil")
+	ErrIndexNeedToBeSaved  = errors.New(`unable to perform action: ImageIndex requires local storage before proceeding.
+	Please use '#Save()' to save the image index locally before attempting this operation`)
+	ErrUnknownMediaType = func(format types.MediaType) error {
+		return fmt.Errorf("unsupported media type encountered in image: '%s'", format)
+	}
+)
+
+type CNBIndex struct {
+	// required
+	v1.ImageIndex // The working Image Index
+
+	// optional
+	Insecure         bool
+	RepoName         string
+	XdgPath          string
+	annotate         Annotate
+	KeyChain         authn.Keychain
+	Format           types.MediaType
+	removedManifests []v1.Hash
+	images           map[v1.Hash]v1.Descriptor
+}
+
+// Annotate a helper struct used for keeping track of changes made to ImageIndex.
```

Comment:
@WYGIN is it possible to provide a little more context on this struct and why it's needed?